solarforecastarbiter/metrics/calculator.py

```diff
@@ -3,98 +3,201 @@
+    normalizer : float
+        Normalized factor (should be in the same units as data). Default is
+        1.0 and only needed for normalized metrics.
+
+    Returns
+    -------
+    pd.DataFrame or dict:
+        Contains all the computed metrics by categories.
+        Structure is:
+
+          * dictionary of forecast 'name' and category types as tuple
+            (e.g., ('Total'), ('Month of the year') )
+          * dictionary with key of metric type
+            (e.g., 'mae', 'rmse')
+          * values of pandas.Series with Index of category values
+            (e.g., pandas.Series([1, 2], index='Aug','Sep'])
+
+    Raises
+    ------
+    RuntimeError
+        If there is no forecast, observation timeseries data or no metrics
+        are specified.
+
+    """
+    calc_metrics = {}
+    calc_metrics['name'] = processed_fx_obs.original.forecast.name
+
+    fx = processed_fx_obs.forecast_values
+    obs = processed_fx_obs.observation_values
+
+    # Check reference forecast is from processed pair, if needed
+    ref_fx = None
+    if any(m in deterministic._REQ_REF_FX for m in metrics):
+        ref_fx = ref_fx_obs.forecast_values
+
+    # No data or metrics
+    if fx.empty:
+        raise RuntimeError("No Forecast timeseries data.")
+    elif obs.empty:
+        raise RuntimeError("No Observation timeseries data.")
+    elif len(metrics) == 0:
+        raise RuntimeError("No metrics specified.")
+
+    # Dataframe for grouping
+    df = pd.concat({'forecast': fx,
+                    'observation': obs,
+                    'reference': ref_fx}, axis=1)
+
+    # Calculate metrics
+    for category in set(categories):
+        calc_metrics[category] = {}
+
+        # total (special category)
+        if category == 'Total':
+            for metric_ in metrics:
+                res = _apply_deterministic_metric_func(
+                    metric_, fx, obs, ref_fx=ref_fx, normalizer=normalizer)
+                calc_metrics[category][metric_] = res
+        else:
+            groupby_category = AVAILABLE_CATEGORIES[category]
+            index_category = getattr(df.index, groupby_category)
+
+            # Calculate each metric
+            for metric_ in metrics:
+
+                metric_values = []
+                cat_values = []
+
+                # Group by category
+                for id, group in df.groupby(index_category):
+
+                    # Calculate
+                    res = _apply_deterministic_metric_func(
+                        metric_, group.forecast, group.observation,
+                        ref_fx=ref_fx, normalizer=normalizer)
+
+                    # Change id of the group
+                    if category == 'Month of the year':
+                        id = calendar.month_abbr[id]
+                    elif category == 'Day of the week':
+                        id = calendar.day_abbr[id]
+
+                    metric_values.append(res)
+                    cat_values.append(id)
```

Comment:
Might be worthwhile to refactor 180-198 into a function like `_calc_metric_for_category(df, index_category, ref_fx, normalizer)`. I get nervous about accumulating values in lists within nested loops and function scopes can help avoid some problems.